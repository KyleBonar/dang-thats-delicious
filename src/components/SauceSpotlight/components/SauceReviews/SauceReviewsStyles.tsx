import styled from "../../../../theme/styled-components";
import Descriptor from "../../../Descriptor/Descriptor";
import { Link } from "../../../Link/Link";

export const StyledContainer = styled.div`
  font-family: AvenirNextReg;
  padding: 1.5rem;
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  margin-bottom: 1.5em;

  @media (min-width: ${props => props.theme.smToMd}) {
    margin-bottom: 3em;
  }
`;
StyledContainer.displayName = "div";

export const StyledLink = styled(Link)`
  > div {
    display: inline;
  }
`;
StyledLink.displayName = "Link";

export const StyledDescriptor = styled(Descriptor)`
  > p {
    margin-top: 0;
    font-style: italic;
  }
`;
StyledDescriptor.displayName = "Descriptor";
