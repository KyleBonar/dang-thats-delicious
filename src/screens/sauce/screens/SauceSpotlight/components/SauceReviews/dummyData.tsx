export const dummyData = [
  {
    _id: "58c6a8aafb4d61115815f8b4",
    date: "2017-03-13T14:11:54.848Z",
    sauce: "5",
    author: "1",
    overall: {
      rating: 4,
      txt:
        "Ive had some umami hot sauces that are too complex in flavor profile and stumble over themselves.... this is not it."
    },
    label: {
      rating: 4,
      txt:
        "Ive had some umami hot sauces that are too complex in flavor profile and stumble over themselves.... this is not it."
    },
    aroma: {
      rating: 4,
      txt:
        "Deep rich smokey aroma with a sweetness of caramalized onions. Really enjoyable on the nose. Not overly complex or simple Deep rich smokey aroma with a sweetness of caramalized onions. Really enjoyable on the nose. Not overly complex or simple "
    },
    taste: { rating: 4, txt: "here is some test text" }
  },
  {
    _id: "58c6a8b8fb4d61115815f8b5",
    date: "2017-03-13T14:12:08.255Z",
    sauce: "5",
    author: "2",
    overall: { rating: 4, txt: "here is some test text" },
    label: { rating: 4, txt: "here is some test text" },
    aroma: { rating: 4, txt: "On the mild side, but balanced well" },
    taste: { rating: 4, txt: "here is some test text" },
    heat: { rating: 4, txt: "On the mild side, but balanced well" },
    note: {
      rating: 4,
      txt:
        "Ive had some umami hot sauces that are too complex in flavor profile and stumble over themselves.... this is not it."
    }
  },
  {
    _id: "58c6a8cefb4d61115815f8b6",
    date: "2017-03-13T14:12:30.578Z",
    sauce: "5",
    author: "3",
    overall: { rating: 4, txt: "On the mild side, but balanced well" },
    label: {
      rating: 4,
      txt:
        "Ive had some umami hot sauces that are too complex in flavor profile and stumble over themselves.... this is not it."
    },
    aroma: { rating: 4, txt: "here is some test text" },
    taste: {
      rating: 4,
      txt:
        "Deep rich smokey aroma with a sweetness of caramalized onions. Really enjoyable on the nose. Not overly complex or simple Deep rich smokey aroma with a sweetness of caramalized onions. Really enjoyable on the nose. Not overly complex or simple "
    },
    heat: { rating: 4, txt: "here is some test text" },
    note: { rating: 4, txt: "On the mild side, but balanced well" }
  },
  {
    _id: "58c6a8e5fb4d61115815f8b7",
    date: "2017-03-13T14:12:53.559Z",
    sauce: "5",
    author: "4",
    overall: { rating: 4, txt: "here is some test text" },
    label: {
      rating: 4,
      txt:
        "Ive had some umami hot sauces that are too complex in flavor profile and stumble over themselves.... this is not it."
    },
    aroma: {
      rating: 4,
      txt:
        "Deep rich smokey aroma with a sweetness of caramalized onions. Really enjoyable on the nose. Not overly complex or simple Deep rich smokey aroma with a sweetness of caramalized onions. Really enjoyable on the nose. Not overly complex or simple "
    },
    taste: {
      rating: 4,
      txt:
        "Ive had some umami hot sauces that are too complex in flavor profile and stumble over themselves.... this is not it."
    },
    heat: {
      rating: 4,
      txt:
        "Deep rich smokey aroma with a sweetness of caramalized onions. Really enjoyable on the nose. Not overly complex or simple Deep rich smokey aroma with a sweetness of caramalized onions. Really enjoyable on the nose. Not overly complex or simple "
    },
    note: { rating: 4, txt: "here is some test text" }
  },
  {
    _id: "58c6a8f6fb4d61115815f8b8",
    date: "2017-03-13T14:13:10.681Z",
    sauce: "5",
    author: "122",
    overall: { rating: 4, txt: "here is some test text" },
    label: {
      rating: 4,
      txt:
        "Ive had some umami hot sauces that are too complex in flavor profile and stumble over themselves.... this is not it."
    },
    aroma: {
      rating: 4,
      txt:
        "Ive had some umami hot sauces that are too complex in flavor profile and stumble over themselves.... this is not it."
    },
    taste: { rating: 4, txt: "here is some test text" },
    heat: { rating: 4, txt: "here is some test text" },
    note: {
      rating: 4,
      txt:
        "Deep rich smokey aroma with a sweetness of caramalized onions. Really enjoyable on the nose. Not overly complex or simple Deep rich smokey aroma with a sweetness of caramalized onions. Really enjoyable on the nose. Not overly complex or simple "
    }
  }
];
