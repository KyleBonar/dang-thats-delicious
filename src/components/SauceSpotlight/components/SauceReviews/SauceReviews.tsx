import * as React from "react";

import SauceReviewBlock from "./components/SauceReviewBlock/SauceReviewBlock";
import { StyledContainer, StyledDescriptor } from "./SauceReviewsStyles";
import { IReview } from "../../../../redux/reviews/types";
import { FlashMessageProps } from "../../../FlashMessage/FlashMessage";

export interface SauceReviewsProps {
  reviews?: IReview[];
  loading: boolean;
  error: FlashMessageProps;
}

const SauceReviews: React.FC<SauceReviewsProps> = props => {
  // defaults
  const _loadingTxt = "loading...";
  const _noReviewsFoundTxt =
    "No reviews found! Have you tried this sauce? Add a review!";

  const { reviews, loading, error } = props;

  if (loading) {
    return <p>{_loadingTxt}</p>;
  }

  if (error.isVisible) {
    return <p>{error.text}</p>;
  }

  if (!reviews || reviews.length === 0) {
    return (
      <StyledContainer>
        <p style={{ marginTop: "0" }}>
          <i>{_noReviewsFoundTxt}</i>
        </p>
      </StyledContainer>
    );
  }

  return (
    <>
      {/* Reviews */}
      <StyledDescriptor title={`Reviews (${reviews ? reviews.length : 0})`}>
        The opinions expressed are solely those of the author.
      </StyledDescriptor>
      <div>
        {reviews.map(review => {
          return <SauceReviewBlock review={review} key={review.reviewID} />;
        })}
      </div>
    </>
  );
};

export default SauceReviews;
