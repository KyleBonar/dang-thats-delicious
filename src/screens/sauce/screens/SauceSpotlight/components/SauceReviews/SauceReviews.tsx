import * as React from "react";
import { connect } from "react-redux";
import { IinitialState } from "../../../../../../redux/configureStore";
import styled from "styled-components";
import { IReview } from "../../../../../../redux/reviews/types";
import { Button } from "../../../../../../components/Button/Button";
import { Link } from "../../../../../../components/Link/Link";
import SauceReviewBlock from "./components/SauceReviewBlock/SauceReviewBlock";
import { dummyData } from "./dummyData";

const StyledContainer = styled.div`
  background-color: ${props => props.theme.formContainerBackgroundColor};
  font-family: AvenirNextReg;
  padding: 1.5rem;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  margin-bottom: 1.5em;

  @media (min-width: ${props => props.theme.smToMd}) {
    margin-bottom: 3em;
  }
`;

export interface SauceReviewsProps {
  id: string; // This is sauce's id that will have to be looked up
  reviews?: IReview[];
  unfoundReviews?: string[];
}

class SauceReviews extends React.Component<SauceReviewsProps, any> {
  public constructor(props: SauceReviewsProps) {
    super(props);
  }

  public render() {
    return (
      <div>
        {this.props.reviews ? (
          this.props.reviews.map(review => {
            return <SauceReviewBlock review={review} key={review._id} />;
          })
        ) : (
          <StyledContainer>
            <p style={{ marginTop: "0" }}>
              <i>No reviews found! Have you tried this sauce? Add a review!</i>
            </p>
            <Link to={`/review/add/?s=${this.props.id}`}>
              <Button>Add Review</Button>
            </Link>
          </StyledContainer>
        )}
      </div>
    );
  }
}

const mapState2Props = (state: IinitialState, ownProps: SauceReviewsProps) => {
  // Make sure we actually have sauces, the specific sauce for the page we are on, and reviews in the state
  const sauces = state.sauces.byId;
  const reviews = state.reviews.byId;
  if (!sauces || !sauces[ownProps.id] || !reviews) {
    return { reviews: dummyData };
  }

  // Grab the array of ids
  const sauceReviews: string[] | undefined = sauces[ownProps.id].reviews;

  // Maybe no reviews on the sauce yet.
  if (!sauceReviews || sauceReviews.length === 0) {
    return {};
  }

  // Push any unfound ids into this arr so we can try, once more, to look for them.
  // This is a small safety net only
  const unfoundReviews: string[] = [];

  // Loop through list of review ids and find the entire review from redux state
  // TODO: Add pagination if there becomes too many reviews on the page
  const reviewsToRender = sauceReviews.map(ReviewID => {
    if (reviews[ReviewID] === undefined) {
      unfoundReviews.push(ReviewID);
    }
    return reviews[ReviewID];
  });

  return { reviews: reviewsToRender, unfoundReviews };
};

export default connect(mapState2Props)(SauceReviews);
