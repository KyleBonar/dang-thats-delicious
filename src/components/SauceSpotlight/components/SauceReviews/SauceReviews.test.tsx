import * as React from "react";
import * as enzyme from "enzyme";

import SauceReviews, { SauceReviewsProps } from "./SauceReviews";
import {
  casual,
  fakeReview,
  ITERATION_SIZE
} from "../../../../utils/testUtils/testUtils";

const fakeSauceReviewBlocks = (): SauceReviewsProps => {
  return {
    reviews: [fakeReview()],
    loading: casual.boolean,
    error: { isVisible: false }
  };
};

describe("<SauceReviews />", () => {
  let wrappers: any = [];
  let props: SauceReviewsProps[] = [];

  beforeAll(() => {
    wrappers = new Array(ITERATION_SIZE).fill(null).map(() => {
      // Generate props
      const fakeProp = fakeSauceReviewBlocks();

      // Add props to collector
      props.push(fakeProp);

      return enzyme.shallow(<SauceReviews {...fakeProp} />);
    });
  });

  it("renders", () => {
    wrappers.forEach(wrapper => {
      expect(wrapper.exists()).toBeTruthy();
    });
  });

  it("matches snapshot", () => {
    wrappers.forEach(wrapper => {
      expect(wrapper.exists()).toMatchSnapshot();
    });
  });
});
