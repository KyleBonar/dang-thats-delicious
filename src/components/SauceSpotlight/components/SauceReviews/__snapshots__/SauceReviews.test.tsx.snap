// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SauceReviews /> matches snapshot 1`] = `true`;

exports[`<SauceReviews /> matches snapshot 2`] = `true`;

exports[`<SauceReviews /> matches snapshot 3`] = `true`;

exports[`<SauceReviews /> matches snapshot 4`] = `true`;

exports[`<SauceReviews /> matches snapshot 5`] = `true`;

exports[`<SauceReviews /> matches snapshot 6`] = `true`;

exports[`<SauceReviews /> matches snapshot 7`] = `true`;

exports[`<SauceReviews /> matches snapshot 8`] = `true`;

exports[`<SauceReviews /> matches snapshot 9`] = `true`;

exports[`<SauceReviews /> matches snapshot 10`] = `true`;

exports[`<SauceReviews /> matches snapshot 11`] = `true`;

exports[`<SauceReviews /> matches snapshot 12`] = `true`;

exports[`<SauceReviews /> matches snapshot 13`] = `true`;

exports[`<SauceReviews /> matches snapshot 14`] = `true`;

exports[`<SauceReviews /> matches snapshot 15`] = `true`;

exports[`<SauceReviews /> matches snapshot 16`] = `true`;

exports[`<SauceReviews /> matches snapshot 17`] = `true`;

exports[`<SauceReviews /> matches snapshot 18`] = `true`;

exports[`<SauceReviews /> matches snapshot 19`] = `true`;

exports[`<SauceReviews /> matches snapshot 20`] = `true`;

exports[`<SauceReviews /> matches snapshot 21`] = `true`;

exports[`<SauceReviews /> matches snapshot 22`] = `true`;

exports[`<SauceReviews /> matches snapshot 23`] = `true`;

exports[`<SauceReviews /> matches snapshot 24`] = `true`;

exports[`<SauceReviews /> matches snapshot 25`] = `true`;

exports[`<SauceReviews /> matches snapshot 26`] = `true`;

exports[`<SauceReviews /> matches snapshot 27`] = `true`;

exports[`<SauceReviews /> matches snapshot 28`] = `true`;

exports[`<SauceReviews /> matches snapshot 29`] = `true`;

exports[`<SauceReviews /> matches snapshot 30`] = `true`;

exports[`<SauceReviews /> matches snapshot 31`] = `true`;

exports[`<SauceReviews /> matches snapshot 32`] = `true`;
